pennylane/hf/tapering.py

```diff
@@ -0,0 +1,277 @@
+        currcol = deepcopy(rref_binary_matrix[:, icol])
+        currcol[irow] = 0
+        rref_binary_matrix[:, icol:] ^= np.outer(currcol, pvtcols)
+
+    return rref_binary_matrix.astype(int)
+
+
+def _kernel(binary_matrix):
+    r"""Computes the kernel of a binary matrix on the binary finite field :math:`\mathbb{Z}_2`.
+
+    Args:
+        binary_matrix (ndarray): binary matrix representation of the Hamiltonian.
+    Returns:
+        nullspace (ndarray): nullspace of the `binary_matrix` where each row correspond to a
+        basis vector in the nullspace.
+
+    **Example**
+
+    .. code-block::
+        >>> binary_matrix = np.array([[1, 0, 0, 0, 0, 1, 0, 0],
+                                      [0, 0, 1, 1, 1, 1, 1, 1],
+                                      [0, 0, 0, 1, 1, 0, 0, 1]])
+        >>> _kernel(binary_matrix)
+         array([[0, 1, 0, 0, 0, 0, 0, 0],
+                [0, 0, 1, 1, 1, 0, 0, 0],
+                [1, 0, 1, 0, 0, 1, 0, 0],
+                [0, 0, 1, 0, 0, 0, 1, 0],
+                [0, 0, 1, 1, 0, 0, 0, 1]])
+
+    """
+
+    # Get the columns with and without pivots
+    pivots = (binary_matrix.T != 0).argmax(axis=0)
+    nonpivots = np.setdiff1d(range(len(binary_matrix[0])), pivots)
+
+    # Initialize the nullspace
+    null_vector = np.zeros((binary_matrix.shape[1], len(nonpivots)), dtype=int)
+    null_vector[nonpivots, np.arange(len(nonpivots))] = 1
+
+    # Fill up the nullspace vectors from the binary matrix
+    null_vector_indices = np.ix_(pivots, np.arange(len(nonpivots)))
+    binary_vector_indices = np.ix_(np.arange(len(pivots)), nonpivots)
+    null_vector[null_vector_indices] = -binary_matrix[binary_vector_indices] % 2
+
+    nullspace = null_vector.T
+    return nullspace
+
+
+def generate_taus(nullspace, num_qubits):
+    r"""Compute the generators :math:`\{\tau_1, \ldots, \tau_k\}` from the nullspace of
+    the binary matrix form of a Hamiltonian over the binary field :math:`\mathbb{Z}_2`.
+    These correspond to the generator set of the :math:`\mathbb{Z}_2`-symmetries present
+    in the Hamiltonian as given in `arXiv:1910.14644 <https://arxiv.org/abs/1910.14644>`_.
+
+    Args:
+        nullspace (list): kernel of the binary matrix corresponding to the Hamiltonian.
+        num_qubits (int): number of wires required to define the Hamiltonian.
+
+    Returns:
+        generators (list): list of generators of symmetries, taus, for the Hamiltonian.
+
+    **Example**
+
+    .. code-block::
+        >>> kernel = np.array([[0, 1, 0, 0, 0, 0, 0, 0],
+                               [0, 0, 1, 1, 1, 0, 0, 0],
+                               [1, 0, 1, 0, 0, 1, 0, 0],
+                               [0, 0, 1, 0, 0, 0, 1, 0],
+                               [0, 0, 1, 1, 0, 0, 0, 1]])
+        >>> generate_taus(kernel, 4)
+         [(1.0) [X1], (1.0) [Z0 X2 X3], (1.0) [X0 Z1 X2], (1.0) [Y2], (1.0) [X2 Y3]]
```

Comment:
Why we need the "1.0"s here? To comply with the OpenFermion input format?

Reply:
This is just an output print statement.